File: climate_preferences.py
"""
Climate preference scoring.

Maps a user's weather preferences (4 axes) onto a 0-100 match score
for any place that has a NASA POWER climate profile.

Preference schema
-----------------
{
    "cold_tolerance":  "dealbreaker" | "tolerable" | "love",
    "heat_tolerance":  "dealbreaker" | "fine"       | "love",
    "rain_tolerance":  "dealbreaker" | "meh"        | "vibe",
    "seasons":         "want_4"      | "mild_ok"    | "want_consistency",
}

Any key can be omitted — omitted axes are excluded from the score.

Score
-----
Each active axis contributes 0-100. Overall is the mean of active axes.
A missing climate profile returns None.
"""

from typing import Optional


def _clamp(val: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, val))


def _linear(val: float, bad: float, good: float) -> float:
    """Map val linearly from bad→0 to good→100, clamped."""
    if good == bad:
        return 50.0
    raw = (val - bad) / (good - bad) * 100.0
    return _clamp(raw, 0.0, 100.0)


def _axis_cold(jan_f: float, pref: str) -> float:
    """Score cold-winter preference. bad=cold (10°F), good=mild (50°F)."""
    if pref == "tolerable":
        return 50.0
    if pref == "dealbreaker":        # hates cold → mild winters score high
        return _linear(jan_f, bad=35.0, good=52.0)
    if pref == "love":               # loves cold → cold winters score high
        return _linear(jan_f, bad=45.0, good=20.0)
    return 50.0


def _axis_heat(jul_f: float, pref: str) -> float:
    """Score summer-heat preference. range: 56°F–93°F."""
    if pref == "fine":
        return 50.0
    if pref == "dealbreaker":        # hates heat → cool summers score high
        return _linear(jul_f, bad=80.0, good=65.0)
    if pref == "love":               # loves heat → hot summers score high
        return _linear(jul_f, bad=70.0, good=85.0)
    return 50.0


def _axis_rain(avg_solar: float, annual_precip_in: float, pref: str) -> float:
    """
    Score rain/grey preference.
    Grey index = 0 (sunny/dry) → 1 (grey/wet), derived from solar + precip.
    """
    if pref == "meh":
        return 50.0
    # Solar: 3.4 (grey) → 5.8 (sunny); normalize to 0-1 sunny
    solar_norm = _clamp((avg_solar - 3.4) / (5.8 - 3.4), 0.0, 1.0)
    # Precip: 6in (dry) → 55in (wet); normalize to 0-1 wet
    precip_norm = _clamp((annual_precip_in - 6.0) / (55.0 - 6.0), 0.0, 1.0)
    # Grey index (higher = greyer, wetter)
    grey = 0.6 * (1.0 - solar_norm) + 0.4 * precip_norm
    if pref == "dealbreaker":        # hates grey/rain → sunny/dry scores high
        return _clamp((1.0 - grey) * 100.0, 0.0, 100.0)
    if pref == "vibe":               # loves grey/rain → grey/wet scores high
        return _clamp(grey * 100.0, 0.0, 100.0)
    return 50.0


def _axis_seasons(swing_f: float, pref: str) -> float:
    """Score seasonal-variation preference. swing = Jul mean - Jan mean."""
    if pref == "mild_ok":
        return 50.0
    if pref == "want_4":             # wants big seasons → high swing scores high
        return _linear(swing_f, bad=18.0, good=40.0)
    if pref == "want_consistency":   # wants year-round consistency → low swing scores high
        return _linear(swing_f, bad=30.0, good=10.0)
    return 50.0


def score_climate_match(
    climate: Optional[dict],
    prefs: dict,
) -> Optional[dict]:
    """
    Score a NASA POWER climate profile against user preferences.

    Args:
        climate: dict with 'months' list (from catalog_climate_profiles.jsonl)
        prefs:   dict with keys cold_tolerance, heat_tolerance, rain_tolerance, seasons

    Returns:
        {score: 0-100, axes: {cold_winter, summer_heat, rain_grey, seasonal}} or None
    """
    if not climate or not climate.get("months"):
        return None

    months = climate["months"]
    by_month = {m["month"]: m for m in months}

    jan_f = by_month.get(1, {}).get("avg_temp_f")
    jul_f = by_month.get(7, {}).get("avg_temp_f")
    if jan_f is None or jul_f is None:
        return None

    annual_precip = sum(m.get("avg_precip_in", 0.0) or 0.0 for m in months)
    solar_vals = [m.get("solar_kwh_m2_day") for m in months if m.get("solar_kwh_m2_day") is not None]
    avg_solar = sum(solar_vals) / len(solar_vals) if solar_vals else 4.5
    swing = jul_f - jan_f

    axes = {}

    if "cold_tolerance" in prefs:
        axes["cold_winter"] = round(_axis_cold(jan_f, prefs["cold_tolerance"]), 1)

    if "heat_tolerance" in prefs:
        axes["summer_heat"] = round(_axis_heat(jul_f, prefs["heat_tolerance"]), 1)

    if "rain_tolerance" in prefs:
        axes["rain_grey"] = round(_axis_rain(avg_solar, annual_precip, prefs["rain_tolerance"]), 1)

    if "seasons" in prefs:
        axes["seasonal"] = round(_axis_seasons(swing, prefs["seasons"]), 1)

    if not axes:
        return None

    overall = round(sum(axes.values()) / len(axes), 1)
    return {"score": overall, "axes": axes}


def load_climate_index(path: str = "data/catalog_climate_profiles.jsonl") -> dict:
    """Load climate profiles into a dict keyed by (name, source)."""
    import json
    index = {}
    with open(path) as f:
        for line in f:
            r = json.loads(line)
            index[(r["name"], r["source"])] = r.get("climate")
    return index

File: frontend/lib/climatePreferences.ts
import type { ClimateIndicators } from './catalogMapTypes'

export type ColdTolerance = 'dealbreaker' | 'tolerable' | 'love'
export type HeatTolerance = 'dealbreaker' | 'fine' | 'love'
export type RainTolerance = 'dealbreaker' | 'meh' | 'vibe'
export type SeasonsPref = 'want_4' | 'mild_ok' | 'want_consistency'

export interface ClimatePreferences {
  cold_tolerance?: ColdTolerance
  heat_tolerance?: HeatTolerance
  rain_tolerance?: RainTolerance
  seasons?: SeasonsPref
}

export interface ClimateMatchResult {
  score: number
  axes: {
    cold_winter?: number
    summer_heat?: number
    rain_grey?: number
    seasonal?: number
  }
}

function clamp(v: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, v))
}

function linear(val: number, bad: number, good: number): number {
  if (good === bad) return 50
  return clamp(((val - bad) / (good - bad)) * 100, 0, 100)
}

function axisCold(jan_f: number, pref: ColdTolerance): number {
  if (pref === 'tolerable') return 50
  if (pref === 'dealbreaker') return linear(jan_f, 35, 52)
  return linear(jan_f, 45, 20) // love cold
}

function axisHeat(jul_f: number, pref: HeatTolerance): number {
  if (pref === 'fine') return 50
  if (pref === 'dealbreaker') return linear(jul_f, 80, 65)
  return linear(jul_f, 70, 85) // love heat
}

function axisRain(avg_solar: number, annual_precip_in: number, pref: RainTolerance): number {
  if (pref === 'meh') return 50
  const solarNorm = clamp((avg_solar - 3.4) / (5.8 - 3.4), 0, 1)
  const precipNorm = clamp((annual_precip_in - 6) / (55 - 6), 0, 1)
  const grey = 0.6 * (1 - solarNorm) + 0.4 * precipNorm
  if (pref === 'dealbreaker') return clamp((1 - grey) * 100, 0, 100)
  return clamp(grey * 100, 0, 100) // vibe
}

function axisSeasons(swing_f: number, pref: SeasonsPref): number {
  if (pref === 'mild_ok') return 50
  if (pref === 'want_4') return linear(swing_f, 18, 40)
  return linear(swing_f, 30, 10) // want consistency
}

export function scoreClimateMatch(
  climate: ClimateIndicators | undefined,
  prefs: ClimatePreferences,
): ClimateMatchResult | null {
  if (!climate) return null

  const axes: ClimateMatchResult['axes'] = {}

  if (prefs.cold_tolerance)
    axes.cold_winter = Math.round(axisCold(climate.jan_f, prefs.cold_tolerance) * 10) / 10
  if (prefs.heat_tolerance)
    axes.summer_heat = Math.round(axisHeat(climate.jul_f, prefs.heat_tolerance) * 10) / 10
  if (prefs.rain_tolerance)
    axes.rain_grey = Math.round(axisRain(climate.avg_solar, climate.annual_precip_in, prefs.rain_tolerance) * 10) / 10
  if (prefs.seasons)
    axes.seasonal = Math.round(axisSeasons(climate.swing_f, prefs.seasons) * 10) / 10

  const vals = Object.values(axes).filter((v): v is number => v !== undefined)
  if (!vals.length) return null

  return {
    score: Math.round((vals.reduce((a, b) => a + b, 0) / vals.length) * 10) / 10,
    axes,
  }
}

export function hasClimatePreferences(prefs: ClimatePreferences): boolean {
  return !!(prefs.cold_tolerance || prefs.heat_tolerance || prefs.rain_tolerance || prefs.seasons)
}
